### PR TITLE
Add a dynamic check for pointer arithmetic of null pointers

### PR DESCRIPTION
**Pull request by mgrang**

### CODE REVIEW

**Comment by mgrang** [comment]
This is a cleaned-up version of https://github.com/microsoft/checkedc-clang/pull/647 with all git merge conflicts resolved.

**Comment by mgrang** [comment]
This patch uncovered a bug in checkedc-parson: https://github.com/microsoft/checkedc-parson/pull/18

File: lib/CodeGen/CGDynamicCheck.cpp [hunk 1]
```diff
@@ -56,21 +56,43 @@ void CodeGenFunction::EmitExplicitDynamicCheck(const Expr *Condition) {
...
 
+void CodeGenFunction::EmitDynamicNonNullCheck(Value *Val,
+                                              const QualType BaseTy) {
+  if (!shouldEmitNonNullCheck(CGM, BaseTy))
+    return;
+
+   ++NumDynamicChecksNonNull;
+
+   Value *ConditionVal = Builder.CreateIsNotNull(Val, "_Dynamic_check.non_null");
+  EmitDynamicCheckBlocks(ConditionVal);	  EmitDynamicCheckBlocks(ConditionVal);
```
Comment (dtarditi) [inline lib/CodeGen/CGDynamicCheck.cpp]: It looks like EmitDynamicCheckBlocks is being called twice here.

File: test/CheckedC/dynamic-checks/array-compound-assign-code-gen.c [hunk 2]
```diff
@@ -1,6 +1,6 @@
 
 // RUN: %clang_cc1 -fcheckedc-extension -verify -verify-ignore-unexpected=note %s -ast-dump | FileCheck %s --check-prefix=CHECK-AST
-// RUN: %clang_cc1 -fcheckedc-extension -verify -verify-ignore-unexpected=note %s -emit-llvm -O0 -o - | FileCheck %s --check-prefix=CHECK-IR
+// RUN: %clang_cc1 -fcheckedc-extension -fno-checkedc-null-ptr-arith -verify -verify-ignore-unexpected=note %s -emit-llvm -O0 -o - | FileCheck %s --check-prefix=CHECK-IR
```
Comment (mgrang) [inline test/CheckedC/dynamic-checks/array-compound-assign-code-gen.c]: By default, null ptr arithmetic checks are enabled. This breaks a few test cases which perform ptr arithmetic but which do not expect the checks. So I had to add the flag `-fno-checkedc-null-ptr-arith` to keep these tests from breaking.

**Comment by mgrang** [comment]
The runtime tests for null ptr arithmetic have been added in https://github.com/microsoft/checkedc/pull/378

**Comment by mgrang** [comment]
> Looks great, thank you!
> 
> Please open a GitHub issue for improving the tests. You and I discuss this in-person. We should update the codegen tests where we are currently disabling null-pointer arithmetic checks. In addition, the tests should cover local variables and parameters.

Done https://github.com/microsoft/checkedc-clang/issues/669.

**Pull request merged**